File: IP/src/iProlog/Spine.java
package iProlog;

/**
 * runtime representation of an immutable list of goals
 * together with top of heap and trail pointers
 * and current clause tried out by head goal
 * as well as registers associated to it
 * <p>
 * note that parts of this immutable lists
 * are shared among alternative branches
 */
class Spine {

    final int hd; // head of the clause to which this corresponds
    final int base; // top of the heap when this was created
    final IntList gs; // goals - with the top one ready to unfold
    final int ttop; // top of the trail when this was created
    int k;
    int[] xs; // index elements
    int[] cs; // array of  clauses known to be unifiable with top goal in gs

    /**
     * creates a spine - as a snapshot of some runtime elements
     */
    Spine(int[] gs0, int base, IntList gs, int ttop, int k, int[] cs) {
        hd = gs0[0];
        this.base = base;
        this.gs = IntList.tail(IntList.app(gs0, gs)); // prepends the goals of clause with head hs
        this.ttop = ttop;
        this.k = k;
        this.cs = cs;
    }
    /**
     * creates a specialized spine returning an answer (with no goals left to solve)
     */
    Spine(int hd, int ttop) {
        this.hd = hd;
        base = 0;
        gs = IntList.empty;
        this.ttop = ttop;

        k = -1;
        cs = null;
    }
}

File: IP/cpp/iProlog/spine.cpp
/*
 * iProlog/C++  [derived from Java version]
 * License: Apache 2.0
 * Copyright (c) 2017 Paul Tarau
 */

#include "spine.h"

namespace iProlog {

    using namespace std;

    /**
     * Creates a spine - as a snapshot of some runtime elements.
     */
    Spine::Spine(
        vector<cell> goal_refs_0, // was gs0/goal_stack_0 [Java]
        int base_0,               // base
        shared_ptr<CellList> goals_0,        // was gs/goal_stack [Java]
        int trail_top_0,
        int k_0,
        vector<int> unifiables_0)
    {
        head = goal_refs_0[0];
        base = base_0;
        trail_top = trail_top_0;
        index_vector = t_index_vector{ -1,-1,-1 };
        last_clause_tried = k_0; 
        goals = CellList::tail(CellList::concat(goal_refs_0, goals_0));
        unifiables = unifiables_0;
    }

    /**
     * "Creates a specialized spine returning an answer (with no goals left to solve)." {Spine.java]
     */
    Spine::Spine(cell h, int tt) {
        head = h;
        base = 0;
        goals = nullptr;
        trail_top = tt;
        last_clause_tried = -1;
        index_vector = { -1,-1,-1 };
    }

} // end namespace
